{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3deb55cf-e477-4d57-a07b-b08c5ae89696",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import Counter\n",
    "from tqdm.notebook import tqdm\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3680dc63-62be-4b7f-8892-7200d2698d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "trans = str.maketrans('ATGC', 'TACG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7a657d2-7547-41c2-9fce-87120cfeb65f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from isotonic_barcode_calling import *\n",
    "import process_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "00017747-e17c-4638-a357-7687f2c7b1a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# minimum score (out of 100) on fuzzy protospacer matching\n",
    "MATCHING_THRESHOLD = 90\n",
    "# number of workers for distance computation in calling\n",
    "NUM_WORKERS = 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a018769d-02fb-43c0-814d-7b12e547ade8",
   "metadata": {},
   "outputs": [],
   "source": [
    "DIRECTORY = \"/data1/normantm/angel/sequencing/AP1_816_puro_virus_output/Samples/DefaultProject\"\n",
    "sampe = \"AP1_816_puro_virus\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c8e1ddd0-ac0d-409d-8f28-993da1ac0bad",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = pd.read_csv(\"120k_barcodes.csv\", header=None)[1].str[30:55].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c6276cb-69b9-4b44-ab53-3836cf217608",
   "metadata": {},
   "outputs": [],
   "source": [
    "K562_separate_sgRNA1 = pd.read_csv(\"twist_K562_816_separate_sgRNA1.csv\", header=None, index_col=0)\n",
    "K562_separate_sgRNA2 = pd.read_csv(\"twist_K562_816_separate_sgRNA2.csv\", header=None, index_col=0)\n",
    "\n",
    "K562_separate_sgRNA1[\"sgRNA1\"] = K562_separate_sgRNA1[1].str[34:54].str.translate(trans).str[::-1]\n",
    "K562_separate_sgRNA2[\"sgRNA2\"] = K562_separate_sgRNA2[1].str[70:90]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7e930db1-3958-4ead-bcb2-6bb048034a88",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sample(sample, separate_sgRNA1, separate_sgRNA2, barcodes, run_name, output_dir):\n",
    "    print(\"reading fastq files\")\n",
    "    umi, _ = process_files.read_fastq(f\"{DIRECTORY}/{sample}/{sample}_UMI_I1.fastq.gz\")\n",
    "    g2 = process_files.read_fastq_g2(f\"{DIRECTORY}/{sample}/{sample}_R2.fastq.gz\")\n",
    "    b = process_files.read_fastq_b(f\"{DIRECTORY}/{sample}/{sample}_R2.fastq.gz\")\n",
    "    g1 = process_files.read_fastq_g1(f\"{DIRECTORY}/{sample}/{sample}_R1.fastq.gz\")\n",
    "    \n",
    "    unique_p1 = np.unique(g1)\n",
    "    unique_p2 = np.unique(g2)\n",
    "\n",
    "    print(\"mapping barcodes\")\n",
    "    barcodes = set(barcodes)\n",
    "\n",
    "    print(\"mapping guides\")\n",
    "    alignments_p1 = chunked_call_alignments_new(unique_p1, separate_sgRNA1.sgRNA1, num_chunks=int(len(unique_p1)/1000), num_workers=NUM_WORKERS)\n",
    "    alignments_p2 = chunked_call_alignments_new(unique_p2, separate_sgRNA2.sgRNA2, num_chunks=int(len(unique_p2)/1000), num_workers=NUM_WORKERS)\n",
    "    alignments_p1.query(\"score >= @MATCHING_THRESHOLD\", inplace=True)\n",
    "    alignments_p2.query(\"score >= @MATCHING_THRESHOLD\", inplace=True)\n",
    "\n",
    "    print(\"filtering reads\")\n",
    "    idx_p1 = np.array([r in alignments_p1.index for r in g1])\n",
    "    idx_p2 = np.array([r in alignments_p2.index for r in g2])\n",
    "    idx_b = np.array([r in barcodes for r in b])\n",
    "    \n",
    "    filtered_idx = idx_p1 & idx_p2 & idx_b\n",
    "    \n",
    "    g1, g2 = g1[filtered_idx], g2[filtered_idx]\n",
    "    b = b[filtered_idx]\n",
    "    umi = umi[filtered_idx]\n",
    "\n",
    "    print(\"writing to files\")\n",
    "    with open(f\"{output_dir}/processed_filtered_reads_{sample}_strict_barcodes.csv\", \"a\") as f:\n",
    "        f.write(\"p1_identity,p2_identity,barcode_mapped,UMI\\n\")\n",
    "    \n",
    "        size = 2000000\n",
    "        chunks = np.arange(0, len(g1), size)\n",
    "        \n",
    "        for i in chunks:\n",
    "            \n",
    "            lines = np.array([alignments_p1.loc[g1[i:i+size]].identity.values,\n",
    "                              alignments_p2.loc[g2[i:i+size]].identity.values,\n",
    "                              b[i:i+size],\n",
    "                              umi[i:i+size]])\n",
    "            \n",
    "            lines = [\",\".join(a)+\"\\n\" for a in lines.T]\n",
    "            \n",
    "            f.writelines(lines)\n",
    "            \n",
    "    with open(f\"{output_dir}/filtered_idx_{sample}.npy\", 'wb') as f:\n",
    "        np.save(f, filtered_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e19266f-0d3c-4b42-996a-72b83f081c9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = \"/data1/normantm/angel/sequencing/AP1_816_puro_virus_output\"\n",
    "run_name = \"20250319_AP1_816_puro_virus\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "15b9291f-51f6-46bf-97c4-96d2d0430b48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading fastq files\n",
      "mapping barcodes\n",
      "mapping guides\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 3128/3128 [11:47<00:00,  4.42it/s]\n",
      "100%|███████████████████████████████████████| 4365/4365 [15:58<00:00,  4.55it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtering reads\n",
      "writing to files\n"
     ]
    }
   ],
   "source": [
    "process_sample(sample, K562_separate_sgRNA1, K562_separate_sgRNA2, barcodes, run_name, output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fffa69f6-d469-4738-a029-80dd7eaeae5c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (angel)",
   "language": "python",
   "name": "angel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
